### PR TITLE
Improve output of integration tests failing check for "output contains string" tests

### PR DESCRIPTION
**Pull request by lptr**
When integration test fails checking for presence of line in output, show only difference with most similar line.

The format is picked explicitly to match IntelliJ's pattern for detecting comparison failures here:

https://github.com/JetBrains/intellij-community/blob/3f97400071e1a0b6a21a6bbd2af1fa27c47467c0/plugins/gradle/java/src/execution/GradleRunnerUtil.java#L97-L116

Output after the changes (17 lines, only non-matching line is shown):

```text
Expected: "result = [b-dir.green, c-dir.red]"
 but: was "result = [b-dir.green, c-dir.green]"
Expected :result = [b-dir.green, c-dir.red]
Actual   :result = [b-dir.green, c-dir.green]
<Click to see difference>


Expected: "result = [b-dir.green, c-dir.red]"
 but: was "result = [b-dir.green, c-dir.green]"
	at app//org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.failOnDifferentLine(OutputScrapingExecutionResult.java:431)
	at app//org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.lambda$assertContentContains$2(OutputScrapingExecutionResult.java:258)
	at java.base@11.0.14/java.util.Optional.ifPresent(Optional.java:183)
	at app//org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertContentContains(OutputScrapingExecutionResult.java:258)
	at app//org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertOutputContains(OutputScrapingExecutionResult.java:267)
	at app//org.gradle.integtests.fixtures.executer.InProcessGradleExecuter$InProcessExecutionResult.assertOutputContains(InProcessGradleExecuter.java:533)
	at app//org.gradle.integtests.fixtures.AbstractIntegrationSpec.outputContains(AbstractIntegrationSpec.groovy:633)
	at org.gradle.integtests.resolve.transform.ArtifactTransformInputArtifactIntegrationTest.honors @PathSensitive(#sensitivity) to input artifact property for project artifact directory when caching(ArtifactTransformInputArtifactIntegrationTest.groovy:534)
```

Output before the changes (65 lines, full build output repeated three times):

```text
Condition failed with Exception:

outputContains("result = [b-dir.green, c-dir.red]")
|
java.lang.AssertionError: Did not find expected text in build output.
Expected: result = [b-dir.green, c-dir.red]
 
Build output:
=======
> Task :b:producer
> Task :c:producer UP-TO-DATE
 
> Task :a:resolve
result = [b-dir.green, c-dir.green]
 
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.failOnMissingOutput(OutputScrapingExecutionResult.java:394)
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertContentContains(OutputScrapingExecutionResult.java:227)
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertOutputContains(OutputScrapingExecutionResult.java:234)
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter$InProcessExecutionResult.assertOutputContains(InProcessGradleExecuter.java:533)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.outputContains(AbstractIntegrationSpec.groovy:633)
	at org.gradle.integtests.resolve.transform.ArtifactTransformInputArtifactIntegrationTest.honors @PathSensitive(#sensitivity) to input artifact property for project artifact directory when caching(ArtifactTransformInputArtifactIntegrationTest.groovy:534)

Condition failed with Exception:

outputContains("result = [b-dir.green, c-dir.red]")
|
java.lang.AssertionError: Did not find expected text in build output.
Expected: result = [b-dir.green, c-dir.red]
 
Build output:
=======
> Task :b:producer
> Task :c:producer UP-TO-DATE
 
> Task :a:resolve
result = [b-dir.green, c-dir.green]
 
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.failOnMissingOutput(OutputScrapingExecutionResult.java:394)
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertContentContains(OutputScrapingExecutionResult.java:227)
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertOutputContains(OutputScrapingExecutionResult.java:234)
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter$InProcessExecutionResult.assertOutputContains(InProcessGradleExecuter.java:533)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.outputContains(AbstractIntegrationSpec.groovy:633)
	at org.gradle.integtests.resolve.transform.ArtifactTransformInputArtifactIntegrationTest.honors @PathSensitive(#sensitivity) to input artifact property for project artifact directory when caching(ArtifactTransformInputArtifactIntegrationTest.groovy:534)

	at org.gradle.integtests.resolve.transform.ArtifactTransformInputArtifactIntegrationTest.honors @PathSensitive(#sensitivity) to input artifact property for project artifact directory when caching(ArtifactTransformInputArtifactIntegrationTest.groovy:534)
Caused by: java.lang.AssertionError: Did not find expected text in build output.
Expected: result = [b-dir.green, c-dir.red]

Build output:
=======
> Task :b:producer
> Task :c:producer UP-TO-DATE

> Task :a:resolve
result = [b-dir.green, c-dir.green]

	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.failOnMissingOutput(OutputScrapingExecutionResult.java:394)
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertContentContains(OutputScrapingExecutionResult.java:227)
	at org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult.assertOutputContains(OutputScrapingExecutionResult.java:234)
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter$InProcessExecutionResult.assertOutputContains(InProcessGradleExecuter.java:533)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.outputContains(AbstractIntegrationSpec.groovy:633)
	... 1 more
```

This is far from perfect, as the new output still shows the expected/actual values three times for some reason in IntelliJ (only twice in TeamCity), but I think it's still a significant improvement in how readable the output is.

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java [hunk 1]
```diff
@@ -220,10 +224,39 @@ public ExecutionResult assertNotOutput(String expectedOutput) {
...
+                    // Measure Levenshtein distance for each line
+                    .map(line -> new LineWithDistance(line, StringUtils.getLevenshteinDistance(line, expectedText)))
+                    // Filter out lines that need more edits than half the length of the line
+                    .filter(pair -> pair.getDistance() < pair.getLine().length() / 2)
+                    // Find the closest match
+                    .min(Comparator.comparingInt(LineWithDistance::getDistance))
+                    .map(LineWithDistance::getLine)
+                    .ifPresent(similarLine -> failOnDifferentLine("Did not find expected text in " + label.toLowerCase() + ", found similar line.", expectedOutput, similarLine));
+            }
             failOnMissingOutput("Did not find expected text in " + label.toLowerCase() + ".", label, expectedOutput, actualText);
```
Comment (tresat) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:260]: Why should this treat single and multi line actual output differently?  Wouldn't the new format also be effective for single-line output?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java [hunk 2]
```diff
@@ -220,10 +224,39 @@ public ExecutionResult assertNotOutput(String expectedOutput) {
...
+    }
+
     @Override
     public ExecutionResult assertContentContains(String actualText, String expectedOutput, String label) {
         String expectedText = LogContent.of(expectedOutput).withNormalizedEol();
         if (!actualText.contains(expectedText)) {
+            if (!expectedText.contains("\n")) {
+                Arrays.stream(actualText.split("\n"))
+                    // Measure Levenshtein distance for each line
+                    .map(line -> new LineWithDistance(line, StringUtils.getLevenshteinDistance(line, expectedText)))
```
Comment (tresat) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:252]: Running Levensthein distance over each line of output text - does this get noticeably slower in wall-clock time when the test has more than a few output lines?  I think this is `O(actual line length * expected line length * num actual lines)`, so it shouldn't, but maybe worth a test against some more substantial output?  Maybe this stream could also be `parallel()`-ized?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java [hunk 3]
```diff
@@ -220,10 +224,39 @@ public ExecutionResult assertNotOutput(String expectedOutput) {
...
+        public int getDistance() {
+            return distance;
+        }
+    }
+
     @Override
     public ExecutionResult assertContentContains(String actualText, String expectedOutput, String label) {
         String expectedText = LogContent.of(expectedOutput).withNormalizedEol();
         if (!actualText.contains(expectedText)) {
+            if (!expectedText.contains("\n")) {
```
Comment (tresat) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:249]: Do we check somewhere that the expected text is not multiline - that wouldn't ever find a match in the actual due to the line splitting.

**Review by tresat — CHANGES_REQUESTED** [review]
I'd consider adding 2 things to the new message:

1) A line number, so that we can quickly find where in long actual output, the mismatch occurred.  Something like:

Expected: "result = [b-dir.green, c-dir.red]"
 but: line 4/26 was "result = [b-dir.green, c-dir.green]"

2) Some sort of disclaimer that this is a best-effort.  I think it's inevitable that at some point this will be used to compare against an actual output where multiple lines are close to the expected line.  It's possible it could identify an actual line which is not the indented matching target.  I can see myself getting confused trying to alter line 4 of the actual output, when really it's line 17 that should match my expectation.

Expected: "result = [b-dir.green, c-dir.red]"
 but: potential match on line 4/26 was "result = [b-dir.green, c-dir.green]"

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java (hunk 3 above)
Comment (lptr) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:249]: I don't understand your question. This is where we check if expected text is single-line for exactly the reason you mention.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java (hunk 1 above)
Comment (lptr) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:260]: For simplicity. For multi-line comparisons the output would look pretty garbled, and falling back to the old format of just displaying the output looks fine to me. We can also improve this further in a new PR if we want to.

Comment (lptr) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:260]: Note that for multi-line outputs we'd need to also compare to similar-sized chunks of the output. Doing so might be worth it, but I don't have the time now to do that, too.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java (hunk 3 above)
Comment (tresat) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:249]: My mistake, sorry.  I was looking for an `else` case and missed that the `failOnMissingOutput` still runs in this situation.

**Review by tresat — APPROVED** [review]

**Comment by lptr** [comment]
@bot-gradle test and merge please

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Pull request merged**